### PR TITLE
Fix issue of initializeRegistry() in IdentityTenantUtil is called for every SSO request

### PR DESCRIPTION
**Pull request by GDRDABARERA**
Resolves wso2/product-is#3076

### CODE REVIEW

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityTenantUtil.java [hunk 1]
```diff
@@ -199,36 +201,40 @@ public static void setBundleContext(BundleContext bundleContext) {
...
-                carbonContext.setTenantDomain(tenantDomain, true);
-                BundleContext bundleContext = IdentityTenantUtil.getBundleContext();
-                if (bundleContext != null) {
-                    ServiceTracker tracker = new ServiceTracker(bundleContext, AuthenticationObserver.class.getName(), null);
-                    tracker.open();
-                    Object[] services = tracker.getServices();
-                    if (services != null) {
-                        for (Object service : services) {
-                            ((AuthenticationObserver) service).startedAuthentication(tenantId);
+        if (tenantIdMap.get(tenantId) == null || Boolean.valueOf(!tenantIdMap.get(tenantId))) {
```
Comment (malithie) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityTenantUtil.java]: What's the reason for this boolean condition
Tenant id is an int value right. Won't this always be false. Why convert for a boolean

Comment (malithie) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityTenantUtil.java]: ! should come before Boolean.valueOf()

Comment (malithie) [reply components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityTenantUtil.java]: Still when invocations happens concurrently there's a chance to get the tenants loaded. Isn't it

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityTenantUtil.java [hunk 2]
```diff
@@ -53,6 +54,7 @@
     private static Log log = LogFactory.getLog(IdentityTenantUtil.class);
     private static TenantRegistryLoader tenantRegistryLoader;
     private static BundleContext bundleContext;
+    protected static ConcurrentHashMap<Integer,Boolean> tenantIdMap = new ConcurrentHashMap<Integer,Boolean>();
```
Comment (Kavindu-Dodan) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityTenantUtil.java]: Improve formatting

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityTenantUtil.java (hunk 1 above)
Comment (Kavindu-Dodan) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityTenantUtil.java]: Can if conditions in line 204 and 206 combined ?

Comment (malithie) [reply components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityTenantUtil.java]: Just noted that the hash map includes a boolean as a value. Why a conversion is needed here

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityTenantUtil.java (hunk 2 above)
Comment (malithie) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityTenantUtil.java]: What's the intention of using a protected variable here

**Comment by malithie** [comment]
What would happen when tenant unloads. 
AFAIK, that would unload tenant registry yet, with this fix as it has been tracked that the tenant has loaded it won't load the tenant again right

**Pull request closed without merging**